### PR TITLE
WIP Annot backend

### PR DESCRIPTION
**Pull request by thumphries**
This is an attempt to revive the Annot backend. This backend serialises type checking information into a textual format that looks as follows:

```
unHtml
Hydrant.Data
23
5
1
unHtml :: Html -> Builder

unAttribute
Hydrant.Data
45
1
1
unAttribute :: Attribute -> (Text, Text)
```

This file format is then consumed by HsColour to add type annotations to highlighted source code. These type annotations appear when moused over. An example can be seen [here](http://goto.ucsd.edu/~rjhala/Annot/Cabal/src/Distribution-Simple-Haddock.html#hscolour).

This work was [largely done by Ranjit Jhala in 2013](https://github.com/ranjitjhala/haddock-annot). It was fully merged into HsColour, but stalled on the way into Haddock because of an unwanted `syb` dependency. Nowadays we have copied large swaths of SYB into Haddock, so integrating this work is easy enough.

While I would rather see this work reimplemented as part of the fancy new hyperlinked backend, I feel the fact that HsColour already supports this format makes completing this backend worthwhile.

### CODE REVIEW

File: haddock-api/src/Haddock/Backends/Annot.hs [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+-- | Get a list of all entities that meet a predicate
+listifyBut :: Typeable r => (r -> Bool) -> [TypeRep] -> GenericQ [r]
+listifyBut p q
+  = everythingButQ (++) q ([] `mkQ` (\x -> if p x then [x] else []))
+
+-- | Types to avoid inspecting. We should not enter any PostTcKind
+-- nor NameSet because these are blank after type checking.
+skipGuards :: [TypeRep]
+skipGuards = [ typeRep (Proxy :: Proxy NameSet)
+             , typeRep (Proxy :: Proxy (PostTc Id Kind))]
```
Comment (thumphries) [inline haddock-api/src/Haddock/Backends/Annot.hs]: I'm not totally sure if this is still necessary - I saw no ill effect from turning these off. The rationale behind them is in Tamar's 2011 blog post linked above.

**Comment by thumphries** [comment]
Most of this code is straight out of the 2011 version. I've cleaned up a few bits that bothered me.

It currently descends into deriving clauses, which seems wrong, I'll try to fix it. Leads to typeclass methods appearing underneath the derivation:

screen shot 2017-11-24 at 9 19 29 am

File: haddock-api/src/Haddock/Backends/Annot.hs [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+findIds a = listifyBut idPredicate skipGuards a
+
+idPredicate :: Var -> Bool
+idPredicate v =
+  isId v && and [
+      not (isDictId v)
+    , not (isGeneratedId v)
+    ]
+
+-- | Filter out compiler-generated names like @$trModule@ or @$cmappend@.
```
Comment (thumphries) [inline haddock-api/src/Haddock/Backends/Annot.hs]: If there's a better way to do this, I'd love to know about it! The `IdDetails` are completely blank for all IDs by the time it gets here.

File: haddock-api/src/Haddock/Backends/Annot.hs [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+  where stripParens ('(':s)  = stripParens s
+        stripParens s        = stripRParens (reverse s)
+        stripRParens (')':s) = stripRParens s
+        stripRParens s       = reverse s
+
+getNames ::  (Data a) => FastString -> a -> [(String, (Id, Maybe Loc))]
+getNames src z = [(unsafePpr x, (x, idLoc src x)) | x <- findIds z]
+
+renderId :: Id -> String
+renderId = showSDocForUser unsafeGlobalDynFlags neverQualify . pprTyThing showSub . AnId
```
Comment (thumphries) [inline haddock-api/src/Haddock/Backends/Annot.hs]: I'm using `unsafeGlobalDynFlags` throughout. Since we run after typechecking this seems safe?

**Comment by alexbiehl** [comment]
Thank you for your effort! This sounds definitely interesting! I will have a look later.

File: haddock-api/src/Haddock/Backends/Annot.hs (hunk 3 above)
Comment (alexbiehl) [reply haddock-api/src/Haddock/Backends/Annot.hs]: Please thread `DynFlags` through. Relying on global state breaks my reasoning.

File: haddock-api/src/Haddock/Backends/Annot.hs (hunk 1 above)
Comment (alexbiehl) [reply haddock-api/src/Haddock/Backends/Annot.hs]: This is ok I guess.

File: haddock-api/src/Haddock/Backends/Annot.hs (hunk 2 above)
Comment (alexbiehl) [reply haddock-api/src/Haddock/Backends/Annot.hs]: Yes, this is ugly but I am not sure there are other ways to check if an `Id` is compiler generated.

File: haddock-api/src/Haddock/Types.hs [hunk 4]
```diff
@@ -151,6 +151,8 @@ data Interface = Interface
     -- | Tokenized source code of module (avaliable if Haddock is invoked with
     -- source generation flag).
   , ifaceTokenizedSrc :: !(Maybe [RichToken])
+
+  , ifaceTcSource :: TypecheckedSource
```
Comment (alexbiehl) [inline haddock-api/src/Haddock/Types.hs]: Let's signal here that `ifaceTcSource` is evaluated.

**Comment by harpocrates** [comment]
FWIW, if the generation of the `AnnMap` moves out of its own module, incorporating this into the hyperlinked backed should be quite easy. In fact, once this gets merged, I don't mind walking through the plumbing of all that myself.

**Comment by thumphries** [comment]
Awesome! Sorry I let this languish, I was busy moving job and country. Will revive the PR this weekend.

**Comment by ntc2** [comment]
This PR was closed accidentally, see [this comment](https://github.com/haskell/haddock/pull/814#issuecomment-387810373) for how to reopen.

**Pull request closed without merging**